src/Components/src/Microsoft.AspNetCore.Components.Server/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,207 @@
+                return segments;
+            }
+        }
+
+        private (IEnumerable<string>, int newPosition) RenderFrames(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            IEnumerable<string> segments = Array.Empty<string>();
+            IEnumerable<string> nextSegments;
+            var nextPosition = position;
+            var endPosition = position + maxElements;
+            while (position < endPosition)
+            {
+                (nextSegments, nextPosition) = RenderCore(frames, position, maxElements);
+                if (position == nextPosition)
+                {
+                    throw new InvalidOperationException("We didn't consume any input.");
+                }
+                position = nextPosition;
+                segments = segments.Concat(nextSegments);
+            }
+
+            return (segments, nextPosition);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderCore(
+            ArrayRange<RenderTreeFrame> frames,
+            int position,
+            int length)
+        {
+            var frame = frames.Array[position];
+            switch (frame.FrameType)
+            {
+                case RenderTreeFrameType.Element:
+                    return RenderElement(frames, position);
+                case RenderTreeFrameType.Attribute:
+                    return RenderAttributes(frames, position, 1);
+                case RenderTreeFrameType.Text:
+                    return (new[] { frame.TextContent }, ++position);
+                case RenderTreeFrameType.Markup:
+                    return (new[] { frame.MarkupContent }, ++position);
+                case RenderTreeFrameType.Component:
+                    return RenderChildComponent(frames, position);
+                case RenderTreeFrameType.Region:
+                    return RenderFrames(frames, position + 1, frame.RegionSubtreeLength - 1);
+                case RenderTreeFrameType.ElementReferenceCapture:
+                case RenderTreeFrameType.ComponentReferenceCapture:
+                    return (Array.Empty<string>(), ++position);
+                default:
+                    throw new InvalidOperationException("Invalid element frame type.");
+            }
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderChildComponent(ArrayRange<RenderTreeFrame> frames, int position)
+        {
+            var frame = frames.Array[position];
+            var childComponentRenderTree = GetCurrentRenderTree(frame.ComponentId);
+            var childFrames = childComponentRenderTree.GetFrames();
+            var (segments, _) = RenderFrames(childFrames, 0, childFrames.Count);
+            return (segments, position + frame.ComponentSubtreeLength);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderElement(ArrayRange<RenderTreeFrame> frames, int position)
```

Comment:
I've used an IEnumerable to represent the sequence of HTML tokens, but will likely change this to a List<string> or some other specialized data structure once we measure the perf.

Reply:
That's interesting. I'm curious about the benefit of this approach versus passing a `StringBuilder` and writing the tokens to it directly. Might that avoid the need to allocate wrapper collections like the arrays on lines 110 and 112?

I'm sure whatever you come up with will be fine, and I agree it's more important to establish the correct behavior + tests right in the first instance. We can tweak the internal representation later if we want.

Reply:
Or maybe instead of a `StringBuilder`, an actual `Stream` (or `TextWriter` or whatever) so it could go directly to the HTTP response. Not sure - you've probably thought this through further than I have.

Reply:
We want something that can be put into an IHtmlContent so that regular Razor can consume it. In the end, (Warning CS buzzword alert) the output type needs only to be a monoid.
StringBuilder with empty string + append.
List with empty list and add
Enumerable with empty sequence and concat.

I need to chat with ryan about it, but I want to avoid "double copy" IE create a string from a string builder or similar only to later copy it again to the regular razor buffer.